civis/io/_tables.py

```diff
@@ -209,8 +215,18 @@ def read_civis_sql(sql, database, use_pandas=False, job_name=None,
     if archive:
         warnings.warn("`archive` is deprecated and will be removed in v2.0.0. "
                       "Use `hidden` instead.", FutureWarning)
+
+    if isinstance(database, str):
+        database = client.get_database_id(database)
+
+    credential_id = credential_id or client.default_credential
+    headers = _get_headers(client, sql, database, credential_id)
```

Comment:
@jmemich,  if I recall correctly you were able to return headers in `civis_to_multifile_csv`.  Do you know how that was implemented?  I'd like to avoid an extra query to retrieve headers if there is some optimization I am unaware of.

Reply:
`civis_to_multifile_csv` uses `force_multifile=True` and returns the json from the manifest file. Civis API writes the headers to the manifest, but only when using the force_multifile option. headers are not provided by the Civis API for other export options.

the challenge with this approach is that it also doesn't combine files so you get back multiple files. we could download those in parallel using threads and then recombine them locally to still return one file to the user.

Reply:
It's possible that there is a side-effect in the query a user passes to `read_civis_sql`.  I.e. a user might pass a query like

```sql
INSERT INTO table VALUES ...;
SELECT * FROM table
```

And the user wouldn't expect that query to run twice (which would happen if we ran a query to get the headers).  In the R client we add `BEGIN READ ONLY` to `read_civis` calls, which we could do here but it would constitute a breaking change.

It's awkward, but I think we need to wrap calls here in `BEGIN READ ONLY`, catch exceptions when the query isn't read only and unload with `PARALLEL OFF` (warning the user) in order to get the headers.

Reply:
Modifying queries has implications for table caching. The read-only queries won't use the same cache as the unmodified queries. That doesn't seem like a big deal here, but it seems worth pointing out.

Reply:
> Modifying queries has implications for table caching.

That's a bummer.  I don't see how we can get around this though, given the problem of possible side effects.  All ideas are welcome on this!

Reply:
we can wrap the query for getting headers in `begin read only;` and if that fails we will proceed with a regular export. since the query for getting headers has a limit 1 it will be returning different results anyway so i don't think we need to worry about table caching here.

Reply:
Oh, right -- I was thinking that the "BEGIN READ ONLY" would be in the regular query as well, but it would only be to get the headers, wouldn't it?